elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +246,261 @@ private Object fetchRelationship(Environment context) {
         }
     }
 
-    /** stub code **/
-    private Object replaceObjects(Environment context) { return null; }
+    /**
+     * handle UPSERT operation
+     * @param context Environment encapsulating graphQL's request environment
+     * @return list of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> upsertObjects(Environment context) {
+        /* sanity check for id and data argument w UPSERT */
+        if(context.ids.isPresent()) {
+            throw new BadRequestException("UPSERT must not include ids");
+        }
+
+        if(!context.data.isPresent()) {
+            throw new BadRequestException("UPSERT must include data argument");
+        }
+
+        Class<?> entityClass;
+        EntityDictionary dictionary = context.requestScope.getDictionary();
+        if(context.isRoot()) {
+            entityClass = dictionary.getEntityClass(context.field.getName());
+        } else {
+            entityClass = dictionary.getParameterizedType(context.parentResource.getResourceClass(), context.field.getName());
+        }
 
-    private Object upsertObjects(Environment context) { return null; }
+        Set<PersistentResource> upsertedObjects = new HashSet();
+        /* upsert */
+        for(Map<String, Object> input : context.data.get()) {
+            upsertedObjects.addAll(graphWalker(input, this::upsertObject, entityClass, context.requestScope));
+        }
+
+        /* fixup relationships */
+        for(Map<String, Object> input : context.data.get()) {
+            graphWalker(input, this::updateRelationship, entityClass, context.requestScope);
+        }
 
-    private Object removeObjects(Environment context) { return null; }
+        /* add relation between parent and nested entity */
+        if(!context.isRoot()) {
+            for(PersistentResource obj : upsertedObjects) {
+                context.parentResource.addRelation(context.field.getName(), obj);
+            }
+        }
+        return upsertedObjects;
+    }
 
-    private Object deleteObjects(Environment context) { return null; }
-}
+    /**
+     * A function to handle upserting (update/create) objects.
+     */
+    @FunctionalInterface
+    private interface Executor<T> {
+        T execute(Entity entity, RequestScope requestScope);
+    }
+
+    /**
+     * Forms the graph from data {@param input} and upsert's {@param function} all the nodes
+     * @param input Input data
+     * @param function Function to process nodes
+     * @param entityClass Entity class
+     * @param requestScope Request scope
+     * @return set of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> graphWalker(Map<String, Object> input, Executor function,
```

Comment:
graphWalker should take an Entity instead of the map.  In general, nothing should deal with the Map except the code that initially converts that to an Entity (that can probably be done in one place).